lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+  double DeltaUsedMemory = computeDelta(OldStat.UsedMemory, NewStat.UsedMemory);
+
+  NewLineInserter nl;
+
+  // Print delta for blocks only if it is above a threshold or we are asked to
+  // dump all changes.
+  if ((SILStatsDumpAll && DeltaBlockCount != 0.0) ||
+      (abs(DeltaBlockCount) > BlockCountDeltaThreshold)) {
+    stats_os() << nl.get();
+    printCounterChange("module", "block", DeltaBlockCount, OldStat.BlockCount,
+                       NewStat.BlockCount, Ctx);
+  }
+
+  // Print delta for instructions only if it is above a threshold or we are
+  // asked to dump all changes.
+  if ((SILStatsDumpAll && DeltaInstCount != 0.0) ||
+      (abs(DeltaInstCount) > InstCountDeltaThreshold)) {
+    stats_os() << nl.get();
+    printCounterChange("module", "inst", DeltaInstCount, OldStat.InstCount,
+                       NewStat.InstCount, Ctx);
+  }
+
+  // Print delta for functions only if it is above a threshold or we are
+  // asked to dump all changes.
+  if ((SILStatsDumpAll && DeltaFunctionCount != 0.0) ||
+      (abs(DeltaFunctionCount) > FunctionCountDeltaThreshold)) {
+    stats_os() << nl.get();
+    printCounterChange("module", "functions", DeltaFunctionCount,
+                       OldStat.FunctionCount, NewStat.FunctionCount, Ctx);
+  }
+
+  // Print delta for the used memory only if it is above a threshold or we are
+  // asked to dump all changes.
+  if ((SILStatsDumpAll && abs(DeltaUsedMemory) > UsedMemoryMinDeltaThreshold) ||
+      (abs(DeltaUsedMemory) > UsedMemoryDeltaThreshold)) {
+    stats_os() << nl.get();
+    printCounterChange("module", "memory", DeltaUsedMemory, OldStat.UsedMemory,
+                       NewStat.UsedMemory, Ctx);
+  }
+
+
+  if (SILStatsDumpAll) {
+    // Dump stats about the number of created and deleted instructions.
+    auto DeltaCreatedInstCount =
+        computeDelta(OldStat.CreatedInstCount, NewStat.CreatedInstCount);
+    auto DeltaDeletedInstCount =
+        computeDelta(OldStat.DeletedInstCount, NewStat.DeletedInstCount);
+    if (DeltaCreatedInstCount != 0.0)
+      printCounterChange("module", "created_inst", DeltaCreatedInstCount,
+                         OldStat.CreatedInstCount, NewStat.CreatedInstCount,
+                         Ctx);
+    if (DeltaDeletedInstCount != 0.0)
+      printCounterChange("module", "deleted_inst", DeltaDeletedInstCount,
+                         OldStat.DeletedInstCount, NewStat.DeletedInstCount,
+                         Ctx);
+  }
+
+  /// Dump collected instruction counts.
+  if (!StatsOnlyInstructionsOptLoc.getInstCountsNum())
+    return;
+
+  for (int i = 0, e = SILInstructionsNum; i < e; ++i) {
+    // Do not print anything, if there is no change.
+    if (OldStat.InstCounts[i] == NewStat.InstCounts[i])
+      continue;
+    SILInstructionKind Kind = SILInstructionKind(i);
+    if (!StatsOnlyInstructionsOptLoc.shouldComputeInstCount(Kind))
+      continue;
+    std::string CounterName = "inst_";
```

Comment:
`llvm::SmallString<64>`? (I forgot if 32 or 64 is the limit), I doubt any name of an instruction is more than 20 characters).